File: Leetcode Daily Challenge/027-September-2025/016-replace-non-coprime-numbers-in-array.java
// https://leetcode.com/problems/replace-non-coprime-numbers-in-array/




import java.util.*;

class Solution {
    // TC : O(N)
    // SC : O(1)
    public List<Integer> replaceNonCoprimes(int[] nums) {
        List<Integer> result = new ArrayList<>();

        for (int num : nums) {
            while (!result.isEmpty()) {
                int prev = result.get(result.size() - 1);
                int gcdVal = getGCD(num, prev);

                if (gcdVal == 1) break;
                result.remove(result.size() - 1);

                int lcmVal = getLCM(num, prev);
                num = lcmVal;
            }
            result.add(num);
        }

        return result;
    }

    private static int getGCD(int a, int b) {
        return (b == 0) ? a : getGCD(b, a % b);
    }

    private static int getLCM(int a, int b) {
        return ((a / getGCD(a, b)) * b);
    }
}

File: Leetcode Daily Challenge/027-September-2025/016-replace-non-coprime-numbers-in-array.cpp
// https://leetcode.com/problems/replace-non-coprime-numbers-in-array/




#include <bits/stdc++.h>
using namespace std;

class Solution {
private:
    using ll = long long;

    ll getLCM(ll a, ll b) {
        return (int)((a / __gcd(a, b)) * b);
    }
public:
    // TC : O(N)
    // SC : O(1)
    vector<int> replaceNonCoprimes(vector<int>& nums) {
        vector<int> result;
        
        for (int &num : nums) {
            while (!result.empty()) {
                int prev   = result.back();
                int gcdVal = __gcd(num, prev);

                if (gcdVal == 1) break;

                result.pop_back();

                int lcmVal = getLCM(num, prev);
                num = lcmVal;
            }
            result.push_back(num);
        }

        return result;
    }
};